docs/source/installation/configuration.md

```diff
@@ -161,10 +161,17 @@ load_utilities:
 
 ## Middleware
 
-`guillotina` is built on `asgi` which provides support for middleware.
+`guillotina` support two types of middlewares:
+
+1. ASGI middlewares: run before the guillotina asgi application and can be used to
+  catch errors or to record timing metrics
```

Comment:
> can be used to catch errors

Right now this is not true because the traversal captures all unhandled exceptions and generates the error response. Ideally, the first ASGI middleware should be the ErrorMiddleware that generates the error response, like in Starlette: https://github.com/encode/starlette/blob/7f8cd041734be3b290cdb178e99c37e1d5a19b41/docs/middleware.md

Agree?

Reply:
Also, I think that guillotina_prometheus should be rewritten as a ASGI middleware because now it runs after the traversal so the response_time/latency isn't accurate (doesn't include the time to run traversal).

Reply:
Cool! That's the reason why our api service it's so fast :)

Reply:
A error handling middleware seems like a nice simple way to provide an example middleware!

Reply:
I'll write it ;) The thing is that we need to decouple this logic from the traversal to middlewares. I'd say that the error and transaction logic should be in separate middlewares.

```
(incoming request) 
ErrorASGIMiddleware
[ User-provided-ASGI-middlewares ... ]
TransactionASGIMiddleware
TraversalRouterMiddleware (creates Request obj)
[ User-provided-middlewares ... ]
View execution
```

Makes sense?

Reply:
I think that should work.

I'm not sure a `TraversalRouterMiddleware` is necessary though?

Reply:
`TraversalRouterMiddleware` is the `AsgiApp` in `guillotina/asgi.py`, so this would be:

```
(incoming request) 
ErrorASGIMiddleware
[ User-provided-ASGI-middlewares ... ]
TransactionASGIMiddleware
GuillotinaAsgiApp (creates Request obj)
[ User-provided-middlewares ... ]
View execution
```

To be sure, I understand that you agree on decoupling the error and transaction logic outside of the Router/Traversal/AsgiApp , right?

Reply:
Yes, error and transaction make sense.

Maybe I just need to see an implementation to understand how I'll like it.

Reply:
Done!